modules/rostests/apitests/advapi32/RegCreateKeyExW.c

```diff
@@ -0,0 +1,158 @@
+    ea[0].Trustee.ptstrName  = (LPTSTR) pEveryoneSID;
+
+    // Create a SID for the BUILTIN\Administrators group.
+    bRes = AllocateAndInitializeSid(&SIDAuthNT, 2,
+        SECURITY_BUILTIN_DOMAIN_RID,
+        DOMAIN_ALIAS_RID_ADMINS,
+        0, 0, 0, 0, 0, 0,
+        &pAdminSID);
+    ok(bRes == TRUE, "AllocateAndInitializeSid Error %ld\n", GetLastError());
+    if (bRes != TRUE) 
+        goto Cleanup;
+
+    // Initialize an EXPLICIT_ACCESS structure for an ACE.
+    // The ACE will allow the Administrators group full access to the key.
+    ea[1].grfAccessPermissions = KEY_ALL_ACCESS;
+    ea[1].grfAccessMode = SET_ACCESS;
+    ea[1].grfInheritance= NO_INHERITANCE;
+    ea[1].Trustee.TrusteeForm = TRUSTEE_IS_SID;
+    ea[1].Trustee.TrusteeType = TRUSTEE_IS_GROUP;
+    ea[1].Trustee.ptstrName  = (LPTSTR) pAdminSID;
+
+    // Create a new ACL that contains the new ACEs.
+    dwRes = SetEntriesInAcl(2, ea, NULL, &pACL);
+    ok(dwRes == ERROR_SUCCESS, "SetEntriesInAcl Error %ld\n", GetLastError());
+    if (dwRes != ERROR_SUCCESS) 
+        goto Cleanup;
+
+    // Initialize a security descriptor.  
+    pSD = (PSECURITY_DESCRIPTOR) LocalAlloc(LPTR, 
+        SECURITY_DESCRIPTOR_MIN_LENGTH); 
+    ok(pSD != NULL, "LocalAlloc Error %ld\n", GetLastError());
+    if (pSD == NULL)
+        goto Cleanup;
+ 
+    bRes = InitializeSecurityDescriptor(pSD, SECURITY_DESCRIPTOR_REVISION); 
+    ok(bRes == TRUE, "InitializeSecurityDescriptor Error %ld\n", GetLastError());
+    if (!bRes)
+        goto Cleanup; 
+
+    // Add the ACL to the security descriptor. 
+    bRes = SetSecurityDescriptorDacl(pSD, 
+        TRUE,     // bDaclPresent flag   
+        pACL, 
+        FALSE);   // not a default DACL 
+    ok(bRes == TRUE, "SetSecurityDescriptorDacl Error %ld\n", GetLastError());
+    if (!bRes)
+        goto Cleanup; 
+
+    // Initialize a security attributes structure.
+    sa.nLength = 0;
+    sa.lpSecurityDescriptor = pSD;
+    sa.bInheritHandle = FALSE;
+
+    // Use the security attributes to set the security descriptor
+    // when you create a key using sa.nLength of 0.
+    sa.nLength = 0;
+    lRes = RegCreateKeyExW(HKEY_CURRENT_USER, L"mykey", 0, L"", 0, 
+            KEY_READ | KEY_WRITE, &sa, &hkSub, &dwDisposition);
+    ok(lRes == ERROR_SUCCESS, "RegCreateKeyExW returned '%ld', expected 0", lRes);
+
+    // Use the security attributes to set the security descriptor
+    // when you create a key using sa.nLength of sa.nLength too short, but not 0.
+    sa.nLength = sizeof(SECURITY_ATTRIBUTES) / 2;
+    lRes = RegCreateKeyExW(HKEY_CURRENT_USER, L"mykey1", 0, L"", 0, 
+            KEY_READ | KEY_WRITE, &sa, &hkSub, &dwDisposition);
+    ok(lRes == ERROR_SUCCESS, "RegCreateKeyExW returned '%ld', expected 0", lRes);
+
+    // Use the security attributes to set the security descriptor
+    // when you create a key using sa.nLength of sa.nLength too long.
+    sa.nLength = sizeof(SECURITY_ATTRIBUTES) + 10;
```

Comment:
"sa" should actually be (at least) that length, to be sure not to access unrelated memory.

Reply:
Since this is an exported function, it is not possible to control the length of sa being at least the sizeof(SECURITY_ATTRIBUTES). So we must deal with any bad cases where it is not here.